testng-core/src/test/java/test/listeners/ListenersTest.java

```diff
@@ -0,0 +1,139 @@
+package test.listeners;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import test.SimpleBaseTest;
+import test.listeners.issue2638.DummyInvokedMethodListener;
+import test.listeners.issue2638.TestClassASample;
+import test.listeners.issue2638.TestClassBSample;
+
+public class ListenersTest extends SimpleBaseTest {
+
+  private static final String[] github2638ExpectedList =
+      new String[] {
+        "test.listeners.issue2638.TestClassASample.testMethod",
+        "test.listeners.issue2638.TestClassBSample.testMethod"
+      };
+
+  @Test(description = "GITHUB-2638", dataProvider = "suiteProvider")
+  public void ensureDuplicateListenersAreNotWiredInAcrossSuites(
+      XmlSuite xmlSuite, Map<String, String[]> expected) {
```

Comment:
Yes. I thought so. But like I mentioned in the other comment, I think we can basically alter the listeners behaviour so that they are executed for the suite for which they are defined. In that case, I think this `expected` can be as a parameter whereas the data provider can be altered.